src/qililab/slurm.py

```diff
@@ -0,0 +1,113 @@
+)
+@argument("-d", "--device", help="Name of the device where you want to execute the SLURM job.")
+@argument(
+    "-l",
+    "--logs",
+    default="./slurm_job_data",
+    help=(f"Path where you want slurm to write the logs for the last {num_jobs_to_keep} jobs"),
+)
+@argument("-n", "--name", default="submitit", help="Name of the slurm job")
+@argument(
+    "-e",
+    "--execution_environment",
+    default=None,
+    help="Select execution environment. Targets slurm by default but if '-e local' the job is run locally.",
+)
+@needs_local_scope
+@register_cell_magic
+def submit_job(line: str, cell: str, local_ns: dict) -> None:
+    """Magic method that queues the content of a cell as a SLURM job.
+
+    WARNING: Variables that start with an underscore (`_`) won't be recognized in the queued job.
+    """
+    # This method does NOT have a standard documentation because it corresponds to an Ipython magic method.
+
+    args = parse_argstring(submit_job, line)
+    output = args.output
+    partition = args.device
+    folder_path = args.logs
+    job_name = args.name
+    execution_env = args.execution_environment
+
+    # Take all the import lines and add them right before the code of the cell (to make sure all the needed libraries
+    # are imported inside the SLURM job)
+    notebook_code = "\n".join(local_ns["In"]).split("\n")
+    import_lines = [line for line in notebook_code if line.startswith(("import ", "from ")) and "slurm" not in line]
+    executable_code = "\n".join(import_lines + [cell])
+
+    # Create the executor that will be used to queue the SLURM job
+    executor = AutoExecutor(folder=folder_path, cluster=execution_env)
+    executor.update_parameters(slurm_partition=partition, name=job_name)
+
+    # Compile the code defined above
+    code = compile(executable_code, "<string>", "exec")
+    # Take all the variables defined by the user in the Jupyter Notebook and the output variable
+    variables = {
+        k: v
+        for k, v in local_ns.items()
+        if not (
+            isinstance(v, ModuleType) or k.startswith("_") or k in {"In", "Out", "exit", "quit", "open", "get_ipython"}
+        )
+    } | {output: None}
+
+    # Define the function that will be queued as a SLURM job
+    def function(code, variables):
+        # Execute the code and return the output variable defined by the user
+        exec(code, variables)
+        return variables[output]
+
+    # Check if output variables are defined or used in the magic cell
+    if not is_variable_used(executable_code, output):
+        raise ValueError(f"Output variable '{output}' was not assigned to any value inside the cell!")
+    # Submit slurm job
+    job = executor.submit(function, code, variables)
+    if execution_env == "local":
+        logger.warning("Your slurm job '%s' with ID %s is running locally, without slurm!", job_name, job.job_id)
+    else:
+        logger.info(f"Your slurm job '{job_name}' with ID {job.job_id} was sent to {partition}!")
```

Comment:
Do we really need to make this distinction? If so, we will not be able to test the second logger line

Reply:
not really